plugins/attention-and-distributed-packing/tests/test_aadp_plugin.py

```diff
@@ -13,22 +13,111 @@
+    """
+    Test framework successfully installs paddingfree plugin
+    """
     with instantiate_framework(
-        read_configuration(CONFIG_PATH_ILAB), require_packages_check=False
+        read_configuration(CONFIG_PATH_PADDINGFREE), require_packages_check=False
     ) as framework:
         for plugin in framework.active_plugins:
             assert isinstance(plugin[1], PaddingFreeAccelerationPlugin)
+
+
+def test_framework_installs_aadp_multipack_and_paddingfree_plugins():
+    """
+    Test framework installs both multipack and paddingfree plugins
+    """
+    pf_config = read_configuration(CONFIG_PATH_PADDINGFREE)
+    mp_config = read_configuration(CONFIG_PATH_MULTIPACK)
+    config = {"training": {**pf_config["training"], **mp_config["training"]}}
+    with instantiate_framework(config, require_packages_check=False) as framework:
+        assert len(framework.active_plugins) == 2
+        for plugin in framework.active_plugins:
+            assert isinstance(
+                plugin[1],
+                (MultipackDataloaderAccelerationPlugin, PaddingFreeAccelerationPlugin),
+            )
+
+
+def test_multipack_sampler_assigns_balanced_tokens():
+    """
+    Ensure that the multipack sampler load balances the tokens amongst the GPUS
+    """
+    num_gpus = 8
+    batch_size_per_device = 32
+    num_samples = 10000
+    seed = 42
+    num_workers = 4
+
+    # 1. Build a test dataset
+    dataset = Dataset.from_list(
+        [
+            {"input_ids": torch.randint(0, 1000, (randint(256, 1024),))}
+            for _ in range(num_samples)
+        ]
+    )
+    lengths = calculate_token_lengths(dataset, num_workers=num_workers)
+
+    # 2.  generate a multipack subset of indices
+    max_batch_len = batch_size_per_device * lengths.mean()
+    tokens_across_rank_multipack = []
+    for rank in range(num_gpus):
+        sampler = MultipackDistributedBatchSampler(
+            batch_max_length=max_batch_len,
+            lengths=lengths,
+            num_replicas=num_gpus,
+            rank=rank,
+            seed=seed,
+            padding=False,
+        )
+        batches = sampler.generate_batches()
+        tokens_across_batches = []
+        for batch in batches:
+            # count all the tokens in the batch
+            num_tokens_across_one_batch = sum(lengths[idx] for idx in batch)
+            tokens_across_batches.append(num_tokens_across_one_batch)
+        # take average number of tokens across the batches
+        average_tokens_across_batches = np.ceil(np.mean(tokens_across_batches))
+        tokens_across_rank_multipack.append(average_tokens_across_batches)
+
+    # 3. generate a random sampled subset of indices
+    tokens_across_rank_random = []
+    perm_indices = torch.randperm(len(dataset)).numpy()
+    # bin indices to respective ranks
+    split_indices_to_ranks = np.array_split(perm_indices, num_gpus)
+    # bin indices in each rank to respective batches
+    split_indices_to_batches = [
```

Comment:
document this so we know `split_indices_to_batches` is a list of what? 

Is it a list with the first dim equal to the num_gpus? document pls